File: java/src/sodium/Node.java
package sodium;

import java.util.HashSet;
import java.util.Set;

public class Node implements Comparable<Node> {
    public final static Node NULL = new Node(Long.MAX_VALUE);

	Node(long rank) {
		this.rank = rank;
	}

	private long rank;
	private Set<Node> listeners = new HashSet<Node>();

	/**
	 * @return true if any changes were made. 
	 */
	boolean linkTo(Node target) {
		if (target == NULL)
			return false;

		boolean changed = target.ensureBiggerThan(rank, new HashSet<Node>());
		listeners.add(target);
		return changed;
	}

	void unlinkTo(Node target) {
		if (target == NULL)
			return;

		listeners.remove(target);
	}

	private boolean ensureBiggerThan(long limit, Set<Node> visited) {
		if (rank > limit || visited.contains(this))
			return false;

		visited.add(this);
		rank = limit + 1;
		for (Node l : listeners)
			l.ensureBiggerThan(rank, visited);
		return true;
	}

	@Override
	public int compareTo(Node o) {
		if (rank < o.rank) return -1;
		if (rank > o.rank) return 1;
		return 0;
	}
}

File: scala/src/main/scala/sodium/Node.scala
package sodium

import scala.collection.mutable.ListBuffer
import scala.ref.WeakReference

class Node(private var rank: Long = 0) extends Comparable[Node] {
  import Node._

  val listeners: ListBuffer[Target] = ListBuffer[Target]()

  /**
    * @return true if any changes were made.
    */
  def linkTo(action: TransactionHandler[Unit], target: Node, outTarget: Array[Target]): Boolean = {
    val changed = target.ensureBiggerThan(rank, Set())
    val t = Target(action, target)
    listeners += t
    outTarget(0) = t
    changed
  }

  def unlinkTo(target: Target): Unit = {
    listeners -= target
  }

  private def ensureBiggerThan(limit: Long, visited: Set[Node]): Boolean =
    if (rank > limit || visited.contains(this)) {
      false
    } else {
      val accVisited = Set(this) ++ visited
      rank = limit + 1
      listeners.foreach(_.node.ensureBiggerThan(rank, accVisited))
      true
    }

  override def compareTo(o: Node): Int = rank.compareTo(o.rank)
}

object Node {
  val NullNode = new Node(Long.MaxValue)

  case class Target(final var action: WeakReference[TransactionHandler[Unit]], final var node: Node)

  object Target {
    def apply(action: TransactionHandler[Unit], node: Node): Target = new Target(new WeakReference(action), node)
  }

}
